[RyuJIT] Emit shlx, sarx, shrx on x64

JulieLeeMSFT opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/41881.
Generates shlx, sarx, shrx for 64 bit shifts if BMI2 platform. 
```
ulong Shlx(ulong x, int y) => x << y;
long Sarx(long x, int y) => x >> y;
ulong Shrx(ulong x, int y) => x >> y;
```
Current codegen:
```
; Method  Test:Shlx(long,int):long
       mov      eax, ecx
       mov      ecx, edx
       shl      eax, cl
       ret      

; Method Test:Sarx(long,int):long
       mov      eax, ecx
       mov      ecx, edx
       sar      eax, cl
       ret      

; Method Test:Shrx(long,int):long
       mov      eax, ecx
       mov      ecx, edx
       shr      eax, cl
       ret   
```
New codegen:
```
for method Test:Shlx(long,int):long
    C4E2E9F7C1           shlx     rax, rcx, rdx

for method Test:Sarx(long,int):long
    C4E2EAF7C1           sarx     rax, rcx, rdx

for method Test:Shrx(long,int):long
    C4E2EBF7C1           shrx     rax, rcx, rdx
```

It needs further work to remove mov when memory address is used instead of all registers (handle it when enabling contained form in https://github.com/dotnet/runtime/issues/67314).
```
ulong ShrxRef(ulong *x, int y) => *x >> y;
    488B01              mov    rax, qword ptr [rcx]
    C4E2EBF7C0          shrx    rax, rax, rdx
```
x86 support needs to be enabled (added it in https://github.com/dotnet/runtime/issues/67314).

JulieLeeMSFT committed:
Emit shlx, sarx, shrx: resolve merge conflict in emitxarch.cpp.

JulieLeeMSFT committed:
Shrx: Swap operand and ShiftBy registers to swap rcx and rdx in correct order.

JulieLeeMSFT committed:
Change only TYP_LONG to Shlx, Sarx and Shrx.

JulieLeeMSFT committed:
Shlx, Sarx, Shrx test cases

JulieLeeMSFT committed:
Shlx, sarx, shrx code clean-up

am11 commented on src/tests/JIT/SIMD/ShiftOperations.csproj:
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
```
```suggestion
    <OutputType>Exe</OutputType>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```

JulieLeeMSFT replied:
Added `<AllowUnsafeBlocks>true</AllowUnsafeBlocks>`.

Wraith2 commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -9493,7 +9507,7 @@ void emitter::emitDispIns(
-            if (ins == INS_bextr || ins == INS_bzhi)
+            if (ins == INS_bextr || ins == INS_bzhi || ins == INS_shrx || ins == INS_shlx || ins == INS_sarx)
             {
                 // BMI bextr and bzhi encodes the reg2 in VEX.vvvv and reg3 in modRM,
```
Does the comment need updating or changing? it specifically mentions the set of things it used to apply to and no longer does.

JulieLeeMSFT replied:
Added 3 shift instructions in the comment.

tannergooding commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -4397,6 +4398,37 @@ void CodeGen::genCodeForShift(GenTree* tree)
         }
     }
+#if defined(TARGET_64BIT)
+    else if (compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2) && tree->OperIs(GT_LSH, GT_RSH, GT_RSZ) &&
```
nit: Should probably use `tree->OperIsShift()`, should also do the `compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2)` check last if possible.

The opportunistic check sets a flag in the R2R metadata and ideally we only set that flag if we know we are in a condition where we would actually use the instruction

JulieLeeMSFT replied:
Got it. Made the changes.

tannergooding commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -4397,6 +4398,37 @@ void CodeGen::genCodeForShift(GenTree* tree)
     }
+#if defined(TARGET_64BIT)
+    else if (compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2) && tree->OperIs(GT_LSH, GT_RSH, GT_RSZ) &&
+             (genActualType(targetType) == TYP_LONG))
```
Why only for `TYP_LONG` and not for 32-bit shifts?

teo-tsirpanis replied:
And we could generate that on x86 as well I guess.

JulieLeeMSFT replied:
Removed the condition and updated LSRA to enable 32 bit shifts.

JulieLeeMSFT replied:
> And we could generate that on x86 as well I guess.

Test passed for x86 as well. Changing the title of this PR.

JulieLeeMSFT replied:
Double checked that it was not enabled for x86. I will not handle x86 optimization in this PR and open an issue as a future work item.

JulieLeeMSFT replied:
Double checked that it was not enabled for x86. I will not handle x86 optimization in this PR and open an issue as a future work item.

tannergooding commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -4397,6 +4398,37 @@ void CodeGen::genCodeForShift(GenTree* tree)
+
+        regNumber shiftByReg = shiftBy->GetRegNum();
+        emitAttr  size       = emitTypeSize(tree);
+        GetEmitter()->emitIns_R_R_R(ins, size, tree->GetRegNum(), shiftByReg, operandReg);
```
Might be worth a note that we don't currently support the contained form due to more complex changes being needed in the emitter.

Ideally we'd fix up the emitter and use `inst_RV_RV_TT` instead so that we can emit `shlx r32a, r/m32, r32b`. Someone would need to walk through the relevant `IF_RWR_RRD_*RD` formats and ensure that it's all handled correctly (noting that technically the format is `IF_RWR_*RD_RRD` but that should be the same as `IF_RWR_RRD_*RD` with swapping op1/op2, like we do for a couple other BMI2 instructions).

JulieLeeMSFT replied:
Opened https://github.com/dotnet/runtime/issues/67314.

kunalspathak replied:
As @tannergooding , lets have a comment about containment and also specify that here, the operands are swapped because the way operands are encoded in these 3 instructions.

JulieLeeMSFT replied:
Added commnets

> As @tannergooding , lets have a comment about containment and also specify that here, the operands are swapped because the way operands are encoded in these 3 instructions.

tannergooding commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -932,6 +932,16 @@ int LinearScan::BuildShiftRotate(GenTree* tree)
         assert(shiftBy->OperIsConst());
     }
+#if defined(TARGET_64BIT)
+    else if (compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2) && tree->OperIs(GT_LSH, GT_RSH, GT_RSZ) &&
```
Same comment as https://github.com/dotnet/runtime/pull/67182/files#r835856687

JulieLeeMSFT replied:
Updated.

JulieLeeMSFT replied:
Done.

tannergooding commented:
Change generally LGTM. Left a suggestion around the ordering of the opportunistic check and potentially logging an issue or a comment around adding support for containment in the future.

JulieLeeMSFT committed:
Emit shlx, sarx, shrx: Enable 32 bit shifts and addressed feedback.

JulieLeeMSFT commented:
> Change generally LGTM. Left a suggestion around the ordering of the opportunistic check and potentially logging an issue or a comment around adding support for containment in the future.

Opened https://github.com/dotnet/runtime/issues/67314.

JulieLeeMSFT commented:
@kunalspathak PTAL.
cc @dotnet/jit-contrib.

kunalspathak commented:
Can you check why there is a regression in coreclr_tests windows/x64?

```
Total bytes of base: 59717535 (overridden on cmd)
Total bytes of diff: 59717754 (overridden on cmd)
Total bytes of delta: 219 (0.00 % of base)
    diff is a regression.
    relative diff is an improvement.
```

kunalspathak commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -4754,7 +4754,7 @@ void Lowering::ContainCheckDivOrMod(GenTreeOp* node)
 {
     assert(node->OperIsShiftOrRotate());
-#ifdef TARGET_X86
+#if defined(TARGET_X86)
```
you could just revert this change...

JulieLeeMSFT replied:
Done.

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -932,6 +932,16 @@ int LinearScan::BuildShiftRotate(GenTree* tree)
     {
         assert(shiftBy->OperIsConst());
     }
+#if defined(TARGET_64BIT)
```
Is this instruction only applicable for x64? I thought it is also valid for x86? @tannergooding ?

JulieLeeMSFT replied:
I wonder why it worked for x86 without changing this part of the code?  

> Is this instruction only applicable for x64? I thought it is also valid for x86? @tannergooding ?

Wraith2 replied:
Isn't `InstructionSet_BMI2` the 32 bit version and for 64 bit you'd want `InstructionSet_BMI2_X64` ? If so and you checked 64 bit correctly elsewhere and then 32 here it would explain it working for 32.

JulieLeeMSFT replied:
Double checked that my change handles only x64 case. Will open an issue to handle x86.

kunalspathak commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -4378,6 +4378,7 @@ void CodeGen::genCodeForShift(GenTree* tree)
             int shiftByValue = (int)shiftBy->AsIntConCommon()->IconValue();
 
 #if defined(TARGET_64BIT)
+
```
delete the extra line.

JulieLeeMSFT replied:
Done.

kunalspathak commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -4397,6 +4398,36 @@ void CodeGen::genCodeForShift(GenTree* tree)
             inst_RV_SH(ins, size, tree->GetRegNum(), shiftByValue);
         }
     }
+#if defined(TARGET_64BIT)
```
same here...why is it only for x64?

JulieLeeMSFT replied:
Handling for x64 only for now.

JulieLeeMSFT replied:
Handling for x64 only for now.

kunalspathak commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -749,6 +749,9 @@ bool emitter::TakesRexWPrefix(instruction ins, emitAttr attr)
             case INS_pdep:
             case INS_pext:
             case INS_rorx:
+            case INS_shlx:
```
As per the changes at other places, we have the code to have it supported only for `TARGET_64BIT`, but here, it is under `TARGET_ARM64`. What is the difference and should it be consistent?

JulieLeeMSFT replied:
It is under #ifdef TARGET_AMD64, not ARM64. So, I guess it is correct.

kunalspathak commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -987,17 +990,25 @@ unsigned emitter::emitOutputRexOrVexPrefixIfNeeded(instruction ins, BYTE* dst, c
                                 case INS_rorx:
                                 case INS_pdep:
                                 case INS_mulx:
+                                case INS_shrx:
```
this code path is also for x86?

JulieLeeMSFT replied:
Added #ifdef TARGET_64BIT.

kunalspathak commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -9493,9 +9507,9 @@ void emitter::emitDispIns(
+            if (ins == INS_bextr || ins == INS_bzhi || ins == INS_shrx || ins == INS_shlx || ins == INS_sarx)
             {
-                // BMI bextr and bzhi encodes the reg2 in VEX.vvvv and reg3 in modRM,
+                // BMI bextr,bzhi, shrx, shlx and sarx encode the reg2 in VEX.vvvv and reg3 in modRM,
```
```suggestion
                // BMI bextr, bzhi, shrx, shlx and sarx encode the reg2 in VEX.vvvv and reg3 in modRM,
```

We need to have similar comment where we swap the operands I pointed out.

JulieLeeMSFT replied:
Added comment.

kunalspathak commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -10288,6 +10302,7 @@ BYTE* emitter::emitOutputAM(BYTE* dst, instrDesc* id, code_t code, CnsVal* addc)
     // For this format, moves do not support a third operand, so we only need to handle the binary ops.
     if (TakesVexPrefix(ins))
     {
+
```
delete

JulieLeeMSFT replied:
Done.

kunalspathak commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -16261,6 +16276,15 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
+        case INS_sarx:
+        case INS_shrx:
+        {
+            result.insLatency    = PERFSCORE_LATENCY_2C;
```
```suggestion
            result.insLatency    += PERFSCORE_LATENCY_1C;
```

It should be similar to that of `rorx` AFAIK.

![image](https://user-images.githubusercontent.com/12488060/160756512-b67f3acf-e807-422d-b721-ec24c7e4e12f.png)

tannergooding replied:
Qhere are you getting that number from? I think that matches the newer hardware and not Skylake, which is what we have used for the other numbers

JulieLeeMSFT replied:
Updated to `result.insLatency    += PERFSCORE_LATENCY_1C;`.

kunalspathak reviewed (changes requested):
Added some comments. I think we need to understand the regression in coreclr/libraries test.

BruceForstall commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,487 @@
+		if (resUInt != expectedUInt)
+		{
+			Console.Write(" != {0} Failed.\n", expectedUInt);
+			return 101;
```
You should consider letting all the tests run, and not exiting on the first failure. Just make sure to return 101 if *any* test fails.

JulieLeeMSFT replied:
Done.

BruceForstall commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -4397,6 +4398,36 @@ void CodeGen::genCodeForShift(GenTree* tree)
+        regNumber shiftByReg = shiftBy->GetRegNum();
+        emitAttr  size       = emitTypeSize(tree);
+        GetEmitter()->emitIns_R_R_R(ins, size, tree->GetRegNum(), shiftByReg, operandReg);
+        genProduceReg(tree);
+
+        return;
```
Delete these lines, and let the normal fall-through make this call and return

JulieLeeMSFT replied:
Done.

JulieLeeMSFT committed:
[shlx/sarx/shrx] Addressed CR feedback.
  - Removed some tests from Arm/Arm64 due to a known undefined behavior
  - Excluded the test from Mono due to a known undefined behavior
  - Resolved merge conflict

JulieLeeMSFT committed:
shlx, sarx, shrx: added ifdef to target x64 only

JulieLeeMSFT commented:
Reran asmdiffs and it is an improvement.

[14:06:59] Summary of Code Size diffs:
[14:06:59] (Lower is better)
[14:06:59]
[14:06:59] Total bytes of base: 129839366 (overridden on cmd)
[14:06:59] Total bytes of diff: 129826617 (overridden on cmd)
[14:06:59] Total bytes of delta: -12749 (-0.01 % of base)
[14:06:59]
[14:06:59]
[14:06:59] 0 total files with Code Size differences (0 improved, 0 regressed), 1125 unchanged.
[14:06:59]
[14:06:59] 0 total methods with Code Size differences (0 improved, 0 regressed), 0 unchanged.

Enabled this only for x64, not x86. Will open a new issue to address it for x86.

JulieLeeMSFT committed:
shlx, sarx, shrx jit-format fix

JulieLeeMSFT commented:
All tests passed. @kunalspathak PTAL.
cc @dotnet/jit-contrib.

kunalspathak commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,501 @@
+using System;
```
Include the license.

kunalspathak commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,501 @@
+            shiftBy = 31;
+            resUInt = Shlx32bit(valUInt, shiftBy);
+            expectedUInt = (uint) (valUInt * Math.Pow(2, (shiftBy % MOD32)));
+            Console.Write("UnitTest Shlx32bit({0},{1}): {2}", valUInt, shiftBy, resUInt);
```
Why not create a `Validate()` function that will do most of these things at one place?

```c#
public int Validate(..., actual, ...) {
  expected = 
  if (expected != actual) { 
    Console.WriteLine("Fail");
    return 101;
  }
  return 100;
}
```

kunalspathak replied:
You can then just have an input array and iterate over it or something like that.

kunalspathak reviewed (changes requested):
Overall looks good. Added some minor suggestions.

kunalspathak commented:
Also, there are some significant regressions. Did you figure out why? Eg. here is from asp.net collection.
 
```
          16 (8.51 % of base) : 6903.dasm - System.Number:ComputeProductApproximation(int,long,long):System.ValueTuple`2[UInt64,UInt64]
          14 (0.87 % of base) : 13765.dasm - V8.Crypto.BigInteger:modPow(V8.Crypto.BigInteger,V8.Crypto.BigInteger):V8.Crypto.BigInteger:this
           9 (1.80 % of base) : 13749.dasm - V8.Crypto.BigInteger:subTo(V8.Crypto.BigInteger,V8.Crypto.BigInteger):this
           8 (0.27 % of base) : 26580.dasm - System.Text.RegularExpressions.Symbolic.SymbolicRegexMatcher`1[UInt64][System.UInt64]:FindEndPositionCapturing(System.ReadOnlySpan`1[Char],int,byref,PerThreadData[UInt64]):int:this
           8 (1.61 % of base) : 13767.dasm - V8.Crypto.BigInteger:addTo(V8.Crypto.BigInteger,V8.Crypto.BigInteger):this
           7 (4.43 % of base) : 15897.dasm - BenchmarksGame.ByteString:GetHashCode():int:this
           7 (0.08 % of base) : 7387.dasm - System.Text.RegularExpressions.RegexInterpreter:TryMatchAtCurrentPosition(System.ReadOnlySpan`1[Char]):bool:this
           6 (0.34 % of base) : 27178.dasm - Microsoft.CodeAnalysis.PEModule:GetTypeAndConstructor(System.Reflection.Metadata.MetadataReader,System.Reflection.Metadata.CustomAttributeHandle,byref,byref):bool
           6 (1.61 % of base) : 17686.dasm - System.Collections.BitArray:set_Length(int):this
           6 (5.45 % of base) : 3915.dasm - System.Text.Encodings.Web.TextEncoderSettings:AllowRange(System.Text.Unicode.UnicodeRange):this
           5 (1.18 % of base) : 16464.dasm - System.Number:AssembleFloatingPointBits(byref,long,int,bool):long
           4 (0.05 % of base) : 28137.dasm - Microsoft.CodeAnalysis.CSharp.Binder:FoldNeverOverflowBinaryOperators(int,Microsoft.CodeAnalysis.ConstantValue,Microsoft.CodeAnalysis.ConstantValue):System.Object
           4 (0.36 % of base) : 27480.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE.PENamedTypeSymbol:MakeDeclaredBaseType():Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol:this
           4 (0.40 % of base) : 27192.dasm - Microsoft.CodeAnalysis.MetadataReaderExtensions:IsTheObjectClass(System.Reflection.Metadata.MetadataReader,System.Reflection.Metadata.TypeDefinition):bool
           4 (4.30 % of base) : 13741.dasm - MontgomeryReducer:.ctor(V8.Crypto.BigInteger):this
           4 (1.53 % of base) : 5738.dasm - System.Net.Http.HPack.IntegerEncoder:Encode(int,int,System.Span`1[Byte],byref):bool
           4 (3.45 % of base) : 16465.dasm - System.Number:RightShiftWithRounding(long,int,bool):long
           4 (4.40 % of base) : 12332.dasm - System.Text.RegularExpressions.Symbolic.BDD:GetMin():long:this
           3 (1.48 % of base) : 1197.dasm - <>c:<EmitMatchCharacterClass>b__122_1(System.Span`1[Char],System.String):this
           3 (0.18 % of base) : 9584.dasm - BenchmarksGame.BinaryTrees_5:Bench(int,bool):int
```

JulieLeeMSFT commented:
> Also, there are some significant regressions. Did you figure out why? Eg. here is from asp.net collection.

They are due to register assignment changes. Bytes regressed in some files but perfscores are better. Some adds NOP of 4 bytes.

## benchmarks.run.windows.x64.checked.mch:
Top method regressions (bytes):
          16 (8.51 % of base) : 6903.dasm  
          14 (0.87 % of base) : 13765.dasm 

- 16 (8.51 % of base) : 6903.dasm:  better perfscore for diff.

Base:
```
       shr      r11, cl
						;; size=13 bbWeight=0.50 PerfScore 1.12
```

Diff:
    

```
       shrx     r11, r11, rdx
						;; size=15 bbWeight=0.50 PerfScore 0.38
```

- 14 (0.87 % of base) : 13765.dasm

Base:
```
       sub      ecx, eax
       sar      edx, cl
       and      edx, dword ptr [rsp+7CH]
       jmp      SHORT G_M1573_IG21
						;; size=40 bbWeight=2    PerfScore 37.50
```

Diff:
```
       mov      edx, ebx
       sub      edx, eax
       sarx     ecx, ecx, edx
       and      ecx, dword ptr [rsp+7CH]
       jmp      G_M1573_IG21
						;; size=46 bbWeight=2    PerfScore 34.50
```

## aspnet.run.windows.x64.checked.mch:
Top method regressions (bytes):
- 31 (11.79 % of base) : 29879.dasm  
  - Checked it and shlx block is 3 bytes larger but `perfScore is 28.71` compared to base `PerfScore 34.65`.
- 19 (8.56 % of base) : 6388.dasm 
Base:
`; Total bytes of code 222, prolog size 14, PerfScore 3321.21, instruction count 69`
Diff
`; Total bytes of code 241, prolog size 14, PerfScore 2796.93, instruction count 69`
- 19 (8.56 % of base) : 5120.dasm 
  -- It add NOP of 4 bytes
Base:
```
G_M35765_IG02:        ; gcrefRegs=000000C0 {rsi rdi}, byrefRegs=00000000 {}, byref, isz
       test     rdi, rdi
       je       SHORT G_M35765_IG09
       xor      ebx, ebx
       mov      ebp, dword ptr [rdi+8]
       test     ebp, ebp
       jle      SHORT G_M35765_IG08
						;; size=18 bbWeight=1    PerfScore 4.75
```

Diff
```
G_M35765_IG02:        ; gcrefRegs=000000C0 {rsi rdi}, byrefRegs=00000000 {}, byref
       test     rdi, rdi
       je       G_M35765_IG09
       xor      ebx, ebx
       mov      ebp, dword ptr [rdi+8]
       test     ebp, ebp
       jle      SHORT G_M35765_IG08
		  ;; NOP compensation instructions of 4 bytes.
						;; size=22 bbWeight=1    PerfScore 4.75
```

JulieLeeMSFT committed:
[shlx/sarx/shrx] Addressed feedback.

BruceForstall commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,253 @@
+    private static ulong Rorx(ulong x) => BitOperations.RotateRight(x, 2);
+
+    public int Validate(actual, expected)
+    {
+        if (expected != actual)
+        {
+            Console.WriteLine("Fail");
+            return 101;
+        }
+        return 100;
+    }
```
Is this version of `Validate` even used?

BruceForstall reviewed (dismissed):
Why is this only for x64, not for x86 as well? Is there significant work to enable it for x86? Shouldn't it just fall out?

JulieLeeMSFT committed:
[Emit shlx, sarx, shrx] addressed feedback on unit test

JulieLeeMSFT commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,253 @@
+                shiftBy = shiftByShlx32bit[idx];
+                resUInt = Shlx32bit(valUInt, shiftBy);
+                expectedUInt = (uint)(valUInt * Math.Pow(2, (shiftBy % MOD32)));
+                returnCode = Validate<uint>(valUInt, shiftBy, resUInt, expectedUInt);
```
Fixed.

JulieLeeMSFT commented:
> Why is this only for x64, not for x86 as well? Is there significant work to enable it for x86? Shouldn't it just fall out?

It was not a simple fall out, so I discussed with Kunal to skip x86 for now.

BruceForstall commented:
> It was not a simple fall out, so I discussed with Kunal to skip x86 for now.

Can you describe what problems were encountered, and what is required to implement it for x86? The tracking issue https://github.com/dotnet/runtime/issues/67314 doesn't have any details.

JulieLeeMSFT committed:
[Emit shlx, sarx, shrx] Removed x86 unit tests.

JulieLeeMSFT commented:
> > It was not a simple fall out, so I discussed with Kunal to skip x86 for now.
> 
> Can you describe what problems were encountered, and what is required to implement it for x86? The tracking issue #67314 doesn't have any details.

It has been a while so I cannot remember exactly, but it was not generating the right code. lsraxarch and emitxarch need to be looked into. Updated #67314.

JulieLeeMSFT commented:
/azp list

JulieLeeMSFT commented:
@kunalspathak and @BruceForstall it is ready to review.

kunalspathak commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,261 @@
+            long expectedLong = 0;
+            int MOD64 = 64;
+
+/* TODO: Enable 32bit test when x86 shift is enabled.
```
I would suggest deleting the commented code.

JulieLeeMSFT committed:
[Emit shlx, sarx, shrx] Remove 32bit test cases.

kunalspathak commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -4803,7 +4803,7 @@ void Lowering::ContainCheckShiftRotate(GenTreeOp* node)
         MakeSrcContained(node, source);
     }
-#endif // !TARGET_X86
+#endif
```
I mean put it back (revert the deletion of the comment `// !TARGET_X86`).

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -925,6 +925,17 @@ int LinearScan::BuildShiftRotate(GenTree* tree)
+    else if (tree->OperIsShift() && !tree->isContained() &&
+             compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2))
+    {
+        // It handles all register forms, but it does not handle contained form for memory operand.
```
```suggestion
        // shlx (as opposed to mov+shl) instructions handles all register forms, but it does not handle contained form for memory operand. Likewise for sarx and shrx.
```

JulieLeeMSFT replied:
Done.

kunalspathak commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,170 @@
+
+        try
+        {
+            ulong valULong = 0;
```
I would define these variables closer to their usage.

JulieLeeMSFT replied:
Fixed.

kunalspathak commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -0,0 +1,170 @@
+            ulong valULong = 0;
+            long valLong = 0;
+            int shiftBy = 0;
+            ulong resULong = 0;
```
Is it worth adding short and ints as test cases?

JulieLeeMSFT replied:
Added.

kunalspathak reviewed (approved):
LGTM...added some suggestions.

JulieLeeMSFT committed:
[Emit shlx, sarx, shrx] Update a comment in lsraxarch.cpp

Co-authored-by: Kunal Pathak <Kunal.Pathak@microsoft.com>

JulieLeeMSFT committed:
[Emit shlx, sarx, shrx] Added uint, int, ushort and short test cases.

kunalspathak commented on src/tests/JIT/SIMD/ShiftOperations.cs:
```diff
@@ -9,19 +9,76 @@
+        switch (x)
+        {
+            case ulong a:
+                ulong resUlong = ((ulong)a) << y;
```
Why doesn't this generic work?

```c#
T res = ((T)x) << y;
return (R)Convert.ChangeType(res,typeof(R));
```

JulieLeeMSFT replied:
shift operators does not work on generics. 
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators
`Because the shift operators are defined only for the int, uint, long, and ulong types, the result of an operation always contains at least 32 bits. If the left-hand operand is of another integral type (sbyte, byte, short, ushort, or char), its value is converted to the int type`

JulieLeeMSFT replied:
@kunalspathak, this is the error message.
`error CS0019: Operator '<<' cannot be applied to operands of type 'T' and 'int'`
All tests passed, so merging it now. Thanks for all the code reviews.

kunalspathak commented:
Improvements in linux/x64 https://github.com/dotnet/runtime/pull/67182
